Feature/undefined expression allowed

rajyan opened the pull request:
fixes https://github.com/phpstan/phpstan/issues/6107
fixes https://github.com/phpstan/phpstan/issues/5971
fixes https://github.com/phpstan/phpstan/issues/5337
fixes https://github.com/phpstan/phpstan/issues/6899

This PR introduces a new method `isUndefinedExpressionAllowed` to `Scope` interface and new property `currentlyAllowedUndefinedExpression`  of `MutatingScope`.

Currently, access to undefined property is allowed in assign. In `Isset_` and `Empty_` it is simulating an assign by `lookForEnterExpressionAssign` to avoid undefined propeties, but it changes from read access to write access which are causing false positives/negatives like issues in above.
The purpose of this feature is to separate `isInExpressionAssign` and undefined property to avoid those issues.
Also, this feature can make allowing/disallowing undefined expressions much more flexibly like https://github.com/phpstan/phpstan-src/pull/1174/commits/59d89a58c3949c2f812df44b2bcb7f7a2fa03367
(this line is added for backwards compatibility that `Coalesce` doesn't allow dynamic properties)
and may be easier to make strict analysis after `#[AllowDynamicProperties]` is introduced in PHP 8.2 :)
https://github.com/phpstan/phpstan-src/pull/1104#issuecomment-1080559786

rajyan committed:
add isUndefinedExpressionAllowed

rajyan committed:
use lookForSetAllowedUndefinedVariable and separate isUndefinedExpressionAllowed from isInExpressionAssign

rajyan committed:
disallow undefined property in coalesce for backwards compatibility

rajyan committed:
fix tests
fixes https://github.com/phpstan/phpstan/issues/6107
fixes https://github.com/phpstan/phpstan/issues/5971

rajyan commented:
memo for failing test https://github.com/pmmp/PocketMine-MP/blob/e0b07ff3087b652407439a29c941f3b66ca92c86/src/pocketmine/Server.php#L2105-L2109

ondrejmirtes commented:
I'm blown away by this, it's awesome! Additionally, I love when people start working on a change X (#1104), realize that changeset Y can be contributed separately (#1174) and once it's merged, they go back to implementing X again :)

I'm ready to merge this, I just want to ask you to add regression tests for the linked issues you know this fixes.

ondrejmirtes commented:
Also - consider basing this on 1.5.x since it can be considered a bugfix for all those issues.

rajyan commented:
> I'm blown away by this, it's awesome! Additionally, I love when people start working on a change X (https://github.com/phpstan/phpstan-src/pull/1104), realize that changeset Y can be contributed separately (https://github.com/phpstan/phpstan-src/pull/1174) and once it's merged, they go back to implementing X again :)

I'm happy to hear your comments! I think I need two more fixes and can finally complete https://github.com/phpstan/phpstan-src/pull/1104 :)

rajyan commented:
> I'm ready to merge this, I just want to ask you to add regression tests for the linked issues you know this fixes.

I'll fix and add tests and do some small refactoring today. I'll mark it ready when it's finished!

rajyan committed:
refactor

rajyan committed:
rename

rajyan committed:
add regression test for https://github.com/phpstan/phpstan/issues/6899

rajyan commented:
The failing integration test is from `json_decode` + `checkExplicitMixed` (property access is not allowed in strict mixed). I'm not sure how to solve this problem as in https://github.com/phpstan/phpstan/issues/1874
but I think this PR https://github.com/phpstan/phpstan-src/pull/993 may solve the problem.

rajyan committed:
add regression test for https://github.com/phpstan/phpstan/issues/5971 and https://github.com/phpstan/phpstan/issues/6107

rajyan commented:
Rebased and added some regression tests. This PR is ready for review!
The failing test is related to this PR, but not sure how to fix.
https://github.com/phpstan/phpstan-src/pull/1174#issuecomment-1088135373

rajyan commented:
I'll fix this one next https://github.com/phpstan/phpstan-src/blob/db78bf599ec32bef307ee2edc7c6917aab6f2f9a/tests/PHPStan/Rules/Properties/AccessPropertiesRuleTest.php#L573-L576 (this feature makes easier to fix the issue :))
and fix `TypeSpecifier` for `Isset_`. Then complete https://github.com/phpstan/phpstan-src/pull/1104 :+1:

ondrejmirtes commented:
Awesome, thank you!
